File: sdk/python/tempmail_sdk/providers/anonymmail.py
"""
Anonymmail 渠道 — https://anonymmail.net
POST JSON API，需要 cookie session
"""

import random
import string
from typing import List

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "anonymmail"
BASE = "https://anonymmail.net"

HEADERS = {
    "Accept": "*/*",
    "Content-Type": "application/x-www-form-urlencoded",
    "Referer": "https://anonymmail.net/",
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"
    ),
}


def _random_username(length: int = 10) -> str:
    """生成随机用户名"""
    return "".join(random.choices(string.ascii_lowercase + string.digits, k=length))


def _fetch_domains() -> List[str]:
    """获取可用域名列表"""
    resp = tm_http.post(f"{BASE}/api/getDomains", headers=HEADERS, timeout=15)
    resp.raise_for_status()
    data = resp.json()
    if not isinstance(data, list):
        return []
    out: List[str] = []
    for item in data:
        if isinstance(item, dict):
            domain = item.get("domain")
            if isinstance(domain, str) and domain.strip():
                out.append(domain.strip())
    return out


def generate_email(channel: str = CHANNEL) -> EmailInfo:
    """创建临时邮箱：先 HEAD 获取 cookie session，再 POST /api/create"""
    # 获取可用域名
    domains = _fetch_domains()
    if not domains:
        raise RuntimeError("anonymmail: no domains available")
    domain = random.choice(domains)
    username = _random_username()
    email = f"{username}@{domain}"

    # HEAD 请求获取 cookie session
    session = tm_http.get_session()
    session.head(f"{BASE}/", headers=HEADERS, timeout=15)

    # POST 创建邮箱
    resp = tm_http.post(
        f"{BASE}/api/create",
        headers=HEADERS,
        data=f"email={email}",
        timeout=15,
    )
    resp.raise_for_status()
    data = resp.json()
    if not isinstance(data, dict) or not data.get("success"):
        raise RuntimeError("anonymmail: create inbox failed")
    addr = data.get("email")
    if not addr or not str(addr).strip():
        raise RuntimeError("anonymmail: missing email in response")
    return EmailInfo(channel=channel, email=str(addr).strip())


def get_emails(email: str) -> List[Email]:
    """获取邮件列表"""
    if not (email or "").strip():
        raise ValueError("anonymmail: empty email")
    e = email.strip()
    resp = tm_http.post(
        f"{BASE}/api/get",
        headers=HEADERS,
        data=f"email={e}",
        timeout=15,
    )
    resp.raise_for_status()
    data = resp.json()
    if not isinstance(data, dict):
        return []
    # 响应格式: {"email@domain": {"created_at": "...", "emails": [...]}}
    inbox = data.get(e)
    if not isinstance(inbox, dict):
        return []
    rows = inbox.get("emails")
    if not isinstance(rows, list):
        return []
    emails: List[Email] = []
    for raw in rows:
        if not isinstance(raw, dict):
            continue
        # 将 token 字段映射为 id
        normalized = dict(raw)
        if "token" in normalized and "id" not in normalized:
            normalized["id"] = normalized.pop("token")
        # 将 body 字段映射为 html（body 包含 HTML 内容）
        if "body" in normalized and "html" not in normalized:
            normalized["html"] = normalized["body"]
            del normalized["body"]
        emails.append(normalize_email(normalized, e))
    return emails

File: sdk/ruby/lib/tempmail_sdk/providers/anonymmail.rb
# frozen_string_literal: true

module TempmailSdk
  module Providers
    # anonymmail.net 渠道实现
    # POST JSON API，需要 cookie session
    module Anonymmail
      CHANNEL = "anonymmail"
      BASE = "https://anonymmail.net"
      HEADERS = {
        "Accept" => "*/*",
        "Content-Type" => "application/x-www-form-urlencoded",
        "Referer" => "https://anonymmail.net/",
        "User-Agent" => "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " \
                        "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"
      }.freeze

      module_function

      def random_username(length = 10)
        chars = ("a".."z").to_a + ("0".."9").to_a
        Array.new(length) { chars.sample }.join
      end

      # 获取可用域名列表
      def fetch_domains
        resp = Http.post("#{BASE}/api/getDomains", headers: HEADERS, timeout: 15)
        resp.raise_for_status
        data = resp.json
        return [] unless data.is_a?(Array)

        data.select { |item| item.is_a?(Hash) && item["domain"].to_s.strip != "" }
            .map { |item| item["domain"].to_s.strip }
      end

      # @return [EmailInfo]
      def generate_email
        domains = fetch_domains
        raise "anonymmail: no domains available" if domains.empty?

        domain = domains.sample
        username = random_username
        email = "#{username}@#{domain}"

        # HEAD 请求获取 cookie session
        Http.get("#{BASE}/", headers: HEADERS, timeout: 15)

        # POST 创建邮箱
        resp = Http.post("#{BASE}/api/create", headers: HEADERS, body: "email=#{email}", timeout: 15)
        resp.raise_for_status
        data = resp.json
        raise "anonymmail: create inbox failed" unless data.is_a?(Hash) && data["success"]

        addr = data["email"].to_s.strip
        raise "anonymmail: missing email in response" if addr.empty?

        EmailInfo.new(channel: CHANNEL, email: addr)
      end

      # @param email [String]
      # @return [Array<Email>]
      def get_emails(email)
        e = email.to_s.strip
        raise "anonymmail: empty email" if e.empty?

        resp = Http.post("#{BASE}/api/get", headers: HEADERS, body: "email=#{e}", timeout: 15)
        resp.raise_for_status
        data = resp.json
        return [] unless data.is_a?(Hash)

        inbox = data[e]
        return [] unless inbox.is_a?(Hash)

        rows = inbox["emails"]
        return [] unless rows.is_a?(Array)

        rows.select { |r| r.is_a?(Hash) }.map do |raw|
          normalized = raw.dup
          # 将 token 字段映射为 id
          if normalized.key?("token") && !normalized.key?("id")
            normalized["id"] = normalized.delete("token")
          end
          # 将 body 字段映射为 html
          if normalized.key?("body") && !normalized.key?("html")
            normalized["html"] = normalized.delete("body")
          end
          Normalize.normalize_email(normalized, e)
        end
      end
    end
  end
end
